### PR TITLE
NMS-12483: Publish AMD64/ARMv7/ARM64 Docker images for Minion

### PR DESCRIPTION
**Pull request by indigo423**
### All Contributors

* [x] Have you read and followed our [Contribution Guidelines](https://github.com/OpenNMS/opennms/blob/develop/CONTRIBUTING.md)?
* [x] Have you [made an issue in the OpenNMS issue tracker](https://issues.opennms.org)?<br>If so, you should:
  1. update the title of this PR to be of the format: `${JIRA-ISSUE-NUMBER}: subject of pull request`
  2. update the JIRA link at the bottom of this comment to refer to the real issue number
  3. prefix your commit messages with the issue number, if possible
* [x] Have you made a comment in that issue which points back to this PR?
* [x] Have you updated the JIRA link at the bottom of this comment to link to your issue?
* [x] If this is a new or updated feature, is there documentation for the new behavior?
* [ ] If this is new code, are there unit and/or integration tests?
* [ ] If this PR targets a `foundation-*` branch, does it avoid changing files in `$OPENNMS_HOME/etc/`?

### Build Minion for AMD64/ARMv7/ARM64 Architecture

Introducing multiarchitecture build for our Minion Docker file. We use Docker buildx experimental features to build the OCI file for multiple architectures. To get distribution independent and reduce the OCI image size we install now from the Minion tarball assembly instead of the RPM files. To support multiple architectures, we have to migrate away from CentOS 8 as a base image which supports just AMD64. We use Ubuntu Eoan instead.

Building the container image:
1. `./assemble.pl -Dopennms.home=/opt/opennms -DskipTests`
2. `cd opennms-container/minion && make && make install`
Image is built and loaded into the Docker context
3. `make clean-all` -> Docker builder is destroyed, tarball in Docker context removed and OCI image deleted

### Reviewer Hints:

* Removed the OCI build for Minion from the RPM build job
* Added a tarball assembly build job which runs in parallel to the DEB and RPM build jobs
* Image is published to a registry for `master | develop | jira/NMS-12483` for AMD64/ARMv7/ARM64 all other branches just AMD64 as OCI file artifact
* We drop support to build the Docker image from RPMs locally or from remote repositories.
* This build does not contain cross-compiled build for JICMP/JICMP6
* We satisfy requirements to support Windows installer (makensis) with a fake version, so we don't have to install and maintain it on our Ubuntu 16 machine environment.

### Todo:

* [x] Change registry organisation from temporary to `opennms`
* [x] Remove jira/NMS-12483 publish from build job

### External References

* JIRA (Issue Tracker): http://issues.opennms.org/browse/NMS-12483

### CODE REVIEW

**Comment by j-white** [comment]
Nice work @indigo423

File: opennms-container/minion/Makefile [hunk 1]
```diff
@@ -0,0 +1,106 @@
...
+.DEFAULT_GOAL := build
+
+VERSION                 := localbuild
+SHELL                   := /bin/bash -o nounset -o pipefail -o errexit
+BUILD_DATE              := $(shell date -u +"%Y-%m-%dT%H:%M:%SZ")
+BASE_IMAGE              := opennms/deploy-base:jre-1.0.0.b31
+DOCKER_CLI_EXPERIMENTAL := enabled
+DOCKERX_INSTANCE        := env-minion-oci
+DOCKER_REGISTRY         := docker.io
+DOCKER_ORG              := no42org
```
Comment (indigo423) [inline opennms-container/minion/Makefile]: ⚠️ Change this to opennms before merge.

File: .circleci/config.yml [hunk 2]
```diff
@@ -437,6 +432,94 @@ jobs:
...
+                ;;
+              develop)
+                make DOCKER_ARCH="linux/amd64,linux/arm/v7,linux/arm64" \
+                     DOCKER_FLAGS=--push \
+                     VERSION="$(../pom2version.py ../../pom.xml)" \
+                     BUILD_NUMBER="${CIRCLE_BUILD_NUM}" \
+                     BUILD_URL="${CIRCLE_BUILD_URL}" \
+                     BUILD_BRANCH="${CIRCLE_BRANCH}"
+                ;;
+              jira/NMS-12483)
```
Comment (indigo423) [inline .circleci/config.yml]: ⚠️ Remove this before merge.

File: .circleci/config.yml [hunk 3]
```diff
@@ -437,6 +432,94 @@ jobs:
...
+            find ./target -name "*.tar.gz" -type f -not -iname '*source*' -exec cp {} ./target/tarballs/opennms.tar.gz \;
+            find ./opennms-assemblies/minion/target -name "*.tar.gz" -type f -not -iname '*source*' -exec cp {} ./target/tarballs/minion.tar.gz \;
+            find ./opennms-assemblies/sentinel/target -name "*.tar.gz" -type f -not -iname '*source*' -exec cp {} ./target/tarballs/sentinel.tar.gz \;
+      - run:
+          name: Build Minion OCI
+          command: |
+            cd opennms-container/minion
+            make DOCKER_TAG="minion:latest" BUILD_NUMBER="${CIRCLE_BUILD_NUM}" BUILD_URL="${CIRCLE_BUILD_URL}" BUILD_BRANCH="${CIRCLE_BRANCH}"
+            case "${CIRCLE_BRANCH}" in
+              master)
```
Comment (j-white) [inline .circleci/config.yml]:
```suggestion
              "master" | "develop")
```

It looks like the code blocks here are the same. We can de-duplicate them with an OR for the case statement?

**Comment by indigo423** [comment]
Just FYI: I've seen this error twice and had to restart the build:

```
[+] Building 0.0s (0/1)                                                         
 => [internal] booting buildkit                                            0.0s
[+] Building 0.0s (1/1) FINISHED                                                
 => ERROR [internal] booting buildkit                                      0.0s
 => => starting container buildx_buildkit_env-minion-oci0                  0.0s
------
 > [internal] booting buildkit:
------
Error: Error response from daemon: container is marked for removal and cannot be started
```

This seems like an underlying Docker daemon flapping issue.

**Comment by j-white** [comment]
> Just FYI: I've seen this error twice and had to restart the build:
> 
> ```
> [+] Building 0.0s (0/1)                                                         
>  => [internal] booting buildkit                                            0.0s
> [+] Building 0.0s (1/1) FINISHED                                                
>  => ERROR [internal] booting buildkit                                      0.0s
>  => => starting container buildx_buildkit_env-minion-oci0                  0.0s
> ------
>  > [internal] booting buildkit:
> ------
> Error: Error response from daemon: container is marked for removal and cannot be started
> ```
> 
> This seems like an underlying Docker daemon flapping issue.

Yeah, we've seen that in other builds too. Haven't tracked it down yet.

File: opennms-container/minion/Dockerfile [hunk 4]
```diff
@@ -1,88 +1,64 @@
...
+COPY ./container-fs/confd/ /opt/minion/confd/
 
-COPY ./rpms /tmp/rpms
+# If you copy from /opt/minion to /opt/minion the permissions are not preserved
+# We would have 755 for minion:root instead of 775 and prevents writing lock files in /opt/minion
+COPY --chown=10001:0 --from=minion-base /opt /opt
 
-SHELL ["/bin/bash", "-c"]
+# Install confd.io configuration files and scripts and ensure they are executable
+COPY ./container-fs/confd/ /opt/minion/confd/
```
Comment (mattixtech) [inline opennms-container/minion/Dockerfile]: Do we need line `45` given line `38`?

Comment (indigo423) [reply opennms-container/minion/Dockerfile]: Thank you! Very good catch slipped through a resolved merge conflict.

**Pull request merged**